edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/commands/KubernetesCrdCommand.cs

```diff
@@ -0,0 +1,263 @@
+                                this.deviceNamespace,
+                                cancellationToken: token);
+                        },
+                        async () => await this.client.CreateNamespacedSecretAsync(newSecret, this.deviceNamespace, cancellationToken: token));
+                    if (v1Secret == null)
+                    {
+                        throw new InvalidIdentityException("Image pull secret was not properly created");
+                    }
+                }
+                catch (Exception ex) when (!ex.IsFatal())
+                {
+                    Events.SecretCreateUpdateFailed(imagePullSecret.Key, ex);
+                }
+            }
+        }
+
+        public async Task ExecuteAsync(CancellationToken token)
+        {
+            string resourceName = this.iotHubHostname + Constants.K8sNameDivider + this.deviceId;
+            string metaApiVersion = Constants.K8sApi + "/" + Constants.K8sApiVersion;
+
+            var modulesList = new List<KubernetesModule<DockerConfig>>();
+            var secrets = new Dictionary<string, ImagePullSecret>();
+            foreach (var runtime in this.runtimeInfo)
+            {
+                foreach (var m in this.modules)
+                {
+                    var combinedConfig = this.combinedConfigProvider.GetCombinedConfig(m, runtime);
+                    CombinedDockerConfig dockerConfig = combinedConfig as CombinedDockerConfig;
+                    if (dockerConfig != null)
+                    {
+                        var combinedModule = new KubernetesModule<DockerConfig>(m)
+                        {
+                            Config = new DockerConfig(dockerConfig.Image, dockerConfig.CreateOptions)
+                        };
+                        modulesList.Add(combinedModule);
+                        dockerConfig.AuthConfig.ForEach(
+                            auth =>
+                            {
+                                var kubernetesAuth = new ImagePullSecret(auth);
+                                secrets[kubernetesAuth.Name] = kubernetesAuth;
+                            });
+                    }
+                    else
+                    {
+                        throw new InvalidModuleException("Cannot convert combined config into KubernetesModule.");
+                    }
+                }
+            }
+
+            Option<EdgeDeploymentDefinition<DockerConfig>> activeDeployment;
+            try
+            {
+                HttpOperationResponse<object> currentDeployment = await this.client.GetNamespacedCustomObjectWithHttpMessagesAsync(
+                    Constants.K8sCrdGroup,
+                    Constants.K8sApiVersion,
+                    this.deviceNamespace,
+                    Constants.K8sCrdPlural,
+                    resourceName,
+                    cancellationToken: token);
+                string body = JsonConvert.SerializeObject(currentDeployment.Body);
```

Comment:
Why are we serializing this object only to deserialize it later (line 156)?

Reply:
We want to convert the existing CRD (which IIRC is a JSONObject returned as object), into an  EdgeDeploymentDefinition, so we can extract information from it, and then use that information to either replace or create the new EdgeDeploymentDefinition.  This is kind of a quirk with the Kubernetes client and the CustomObject functions.